From d8be9fa82a9c10a49e519db019472b7fe000ce2d Mon Sep 17 00:00:00 2001
From: jjacky <i.am.jack.mail@gmail.com>
Date: Wed, 5 Jun 2013 16:07:40 +0200
Subject: [PATCH] GtkTreeView: Fix issues when removing a column

When removing a column from a treeview, _gtk_tree_view_column_unset_tree_view()
gets called. This would use gtk_container_remove() to remove the button (from
column header) from the treeview.

In addition to a memory leak (the button never being unreferenced), this would
cause quite a few issues in the treeview, where wrong window/button would be
used on remaining columns, going from visual/UI glitches to segfaults.

This is because the button wasn't added to the treeview via GtkContainer, the
treeview was set as parent of the button (via gtk_widget_set_parent() from
gtk_tree_view_column_create_button()). Therefore, the proper way to "undo" this
is call gtk_widget_unparent() on the button.

Signed-off-by: Olivier Brunel <jjk@jjacky.com>
---
 gtk/gtktreeviewcolumn.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gtk/gtktreeviewcolumn.c b/gtk/gtktreeviewcolumn.c
index e1846d5..90696ff 100644
--- a/gtk/gtktreeviewcolumn.c
+++ b/gtk/gtktreeviewcolumn.c
@@ -1459,7 +1459,7 @@ _gtk_tree_view_column_unset_tree_view (GtkTreeViewColumn *column)
 
   if (priv->tree_view && priv->button)
     {
-      gtk_container_remove (GTK_CONTAINER (priv->tree_view), priv->button);
+      gtk_widget_unparent (priv->button);
     }
   if (priv->property_changed_signal)
     {
-- 
1.8.4

